src/tribol/mesh/MfemData.cpp

```diff
@@ -0,0 +1,876 @@
+// Copyright (c) 2017-2023, Lawrence Livermore National Security, LLC and
+// other Tribol Project Developers. See the top-level LICENSE file for details.
+//
+// SPDX-License-Identifier: (MIT)
+
+#include "tribol/mesh/MfemData.hpp"
+
+#include "axom/slic.hpp"
+
+namespace tribol
+{
+
+SubmeshLORTransfer::SubmeshLORTransfer(
+  mfem::ParFiniteElementSpace& ho_submesh_fes,
+  mfem::ParMesh& lor_submesh
+)
+: lor_gridfn_ { CreateLORGridFunction(
+    lor_submesh,
+    std::make_unique<mfem::H1_FECollection>(1, lor_submesh.SpaceDimension()),
+    ho_submesh_fes.GetVDim()
+  ) },
+  lor_xfer_ { ho_submesh_fes, *lor_gridfn_.ParFESpace() }
+{
+  SLIC_WARNING_ROOT("LOR support is experimental at this time.");
+}
+
+void SubmeshLORTransfer::TransferToLORGridFn(
+  const mfem::ParGridFunction& ho_src
+)
+{
+  lor_xfer_.ForwardOperator().Mult(ho_src, lor_gridfn_);
+}
+
+void SubmeshLORTransfer::TransferFromLORGridFn(
+  mfem::ParGridFunction& ho_dst
+) const
+{
+  lor_xfer_.ForwardOperator().MultTranspose(lor_gridfn_, ho_dst);
+}
+
+mfem::ParGridFunction SubmeshLORTransfer::CreateLORGridFunction(
+  mfem::ParMesh& lor_mesh,
+  std::unique_ptr<mfem::FiniteElementCollection> lor_fec,
+  integer vdim
+)
+{
+  mfem::ParGridFunction lor_gridfn {
+    new mfem::ParFiniteElementSpace(
+      &lor_mesh,
+      lor_fec.get(),
+      vdim,
+      mfem::Ordering::byNODES
+    )
+  };
+  lor_gridfn.MakeOwner(lor_fec.release());
+  return lor_gridfn;
+}
+
+SubmeshRedecompTransfer::SubmeshRedecompTransfer(
+  mfem::ParFiniteElementSpace& submesh_fes,
+  SubmeshLORTransfer* lor_xfer,
+  redecomp::RedecompMesh& redecomp
+)
+: submesh_fes_ { submesh_fes },
+  redecomp_fes_ { lor_xfer ?
+    CreateRedecompFESpace(redecomp, *lor_xfer->GetLORGridFn().ParFESpace()) :
+    CreateRedecompFESpace(redecomp, submesh_fes_) },
+  lor_xfer_ { lor_xfer },
+  redecomp_xfer_ { } // default (element transfer) constructor
+{
+  // make sure submesh_fes is a submesh and redecomp's parent is submesh_fes's
+  // submesh
+  SLIC_ERROR_ROOT_IF(
+    !mfem::ParSubMesh::IsParSubMesh(submesh_fes_.GetParMesh()),
+    "submesh_fes must be on a ParSubMesh."
+  );
+  SLIC_ERROR_ROOT_IF(
+    !lor_xfer &&
+      &redecomp.getParent() != submesh_fes_.GetParMesh(),
+    "redecomp's parent must match the submesh_fes ParMesh."
+  );
+  SLIC_ERROR_ROOT_IF(
+    lor_xfer &&
+      &redecomp.getParent() != lor_xfer->GetLORGridFn().ParFESpace()->GetParMesh(),
+    "redecomp's parent must match the submesh_fes ParMesh."
+  );
+}
+
+mfem::GridFunction SubmeshRedecompTransfer::SubmeshToRedecomp(
```

Comment:
The function name makes it sounds like a submesh is used to create a redecomp mesh or somehow a mesh to mesh relation, but then I see that it returns a grid function. Should this be named SubmeshGFToRedecompMeshGF() or something like that?

Reply:
We can update this, but I think the function parameters and the documentation provide enough clarity that this is for transferring grid functions.  If we do change it, I'd prefer something like `GridFn`, since that is less ambiguous than `GF`.